runtime/array.c

```diff
@@ -414,39 +370,20 @@ static value caml_array_gather(intnat num_arrays,
     CAMLassert(pos == size);
   }
 #endif
-  else if (size <= Max_young_wosize) {
+  else if (size > Max_wosize) {
+    /* Array of values, too big. */
+    caml_invalid_argument("Array.concat");
+  }
+  else {
     /* Array of values, small enough to fit in young generation.
        We can use memcpy directly. */
     res = caml_alloc_small(size, 0);
     for (i = 0, pos = 0; i < num_arrays; i++) {
-      memcpy(&Field(res, pos),
-             &Field(arrays[i], offsets[i]),
-             lengths[i] * sizeof(value));
+      caml_blit_fields(arrays[i], offsets[i], res, pos, lengths[i]);
```

Comment:
Once the allocation issue (https://github.com/ocaml/ocaml/pull/10831/files#r776928176) is fixed, care must be taken not to call `caml_blit_fields` on the uninitialised array allocated in the major heap. Since `caml_blit_fields` uses `caml_modify` which in turn uses the write barrier, we may try to darken the uninitialised field and crash.

Reply:
Fixed by https://github.com/ocaml-multicore/ocaml-multicore/pull/822.